Notify users through sns (email only)

dgcaccam opened the pull request:
should allow users to subscribe and get email notifications from any publish to their subscribed topic...

dgcaccam committed:
sns service class to publish and subscribe to topics

dgcaccam committed:
changes to tag and user entities to store topic arns (subscriptions)
changes to data handler and subscription services to work with sns

** a user subscribes to a topic and their email will get attached. They
will get emails with every future publish to that topic. They can also
unsubscribe at any time**

topnessman commented on src/main/java/edu/uwaterloo/ece658/entity/User.java:
```diff
@@ -46,6 +48,9 @@
     private String gender;
 
+    @ElementCollection
+    private Map<String, String> subscriptionArns;
```
Why having this map?

dgcaccam replied:
The way the arns work. When you create a topic you get a topicArn. When you subscribe to a topic the user gets a subscriptionArn. I wanted this so i could keep track of which topicArn the users subscriptionArn mapped to

topnessman commented on src/main/java/edu/uwaterloo/ece658/service/DataHandlerService.java:
```diff
@@ -54,11 +57,14 @@ private Photo createNewPhoto(
         return photo;
     }
-
+    
```
remove space

topnessman commented on src/main/java/edu/uwaterloo/ece658/service/SubscriptionService.java:
```diff
@@ -26,17 +28,21 @@
         tagFacade.edit(tag);
         userFacade.edit(user);
-
+        
```
remove space

topnessman commented on src/main/java/edu/uwaterloo/ece658/service/SubscriptionService.java:
```diff
@@ -47,35 +53,23 @@ public boolean unsubscribeFromTag(String username, String tagName) {
         tagFacade.edit(tag);
         userFacade.edit(user);
-
+        
```
remove space

topnessman commented on src/main/java/edu/uwaterloo/ece658/service/SnsService.java:
```diff
@@ -0,0 +1,77 @@
+    
+    public void createNewTopic(Tag tag) {
+        CreateTopicRequest createTopicRequest =
+                new CreateTopicRequest("MyNewTopic");
```
Why hardcoding "MyNewTopic"?

dgcaccam replied:
supposed to be tagname. I'll fix

topnessman commented on src/main/java/edu/uwaterloo/ece658/service/SnsService.java:
```diff
@@ -0,0 +1,77 @@
+        tag.setTopicArn(createTopicResult.getTopicArn());
+    }
+    
+    public void subscribeToTopic(String topicArn, String tagName, User user) {
```
If a user is subscribed to a tag, he/she implicitly subscribed to the topic(I suppose tag and topic are one-to-one relation?). Why having this subscribeToTopic method?

topnessman replied:
And who is the client for SNSService?

dgcaccam replied:
client is same as the s3 service. When they subscribe to a tag it automatically subscribes them to the topic. It's in the same process

topnessman commented on src/main/java/edu/uwaterloo/ece658/service/SnsService.java:
```diff
@@ -0,0 +1,77 @@
+        user.addSubscriptionArn(tagName, result.getSubscriptionArn());
+    }
+    
+    public void unsubscribeFromTopic(String tagName, User user) {    
```
unsubscribeFromTopic accepts two arguments but subscribeToTopic takes three, why?

dgcaccam replied:
subscribing needs the topicArn while unsubscribing doesn't. unsubscribing is done with the user's subscriptionArn

dgcaccam committed:
small fixes

dgcaccam committed:
- change of region for SNS
- updated unsubscribe logic (cannot store subsciptionarn as its not 
	generated until after the email confirmation. Amazon SNS has a method
	to find the subscriptionArns at runtime)
- SNS logic tested in separate project
